src/common/snippets/src/lowered/pass/mark_invariant_shape_path.cpp

```diff
@@ -0,0 +1,124 @@
+           ov::is_type<ov::snippets::op::ConvertTruncation>(node) ||
+           ov::is_type<ov::snippets::op::ConvertSaturation>(node);
+}
+static bool is_scalar_op(const ExpressionPtr& expr) {
+    const auto& node = expr->get_node();
+    return ov::is_type<ov::snippets::op::HorizonMax>(node) ||
+           ov::is_type<ov::snippets::op::HorizonSum>(node) ||
+           ov::is_type<ov::snippets::op::VectorBuffer>(node) ||
+           ov::is_type<ov::snippets::op::BroadcastMove>(node);
+}
+}  // namespace
+
+size_t MarkInvariantShapePath::getInvariantPortShapePath(const ExpressionPort& port) {
+    auto& rt = get_rt_info(port);
+    const auto rinfo = rt.find("InvariantShapePath");
+    OPENVINO_ASSERT(rinfo != rt.end(), "Invariant path for this expression port has not been marked!");
+    return rinfo->second.as<size_t>();
+}
+
+void MarkInvariantShapePath::SetInvariantPortShapePath(const ExpressionPort& port, size_t value) {
+    auto& rt = get_rt_info(port);
+    rt["InvariantShapePath"] = value;
+}
+
+ov::RTMap& MarkInvariantShapePath::get_rt_info(const ExpressionPort& port) {
+    const auto& node = port.get_expr()->get_node();
+    const auto port_idx = port.get_index();
+    const auto is_input = port.get_type() == ExpressionPort::Input;
+    OPENVINO_ASSERT((is_input && (port_idx < node->get_input_size())) || (!is_input && (port_idx < node->get_output_size())),
+                    "Node has incompatible port count with the expression");
+    return is_input ? node->input(port_idx).get_rt_info() : node->output(port_idx).get_rt_info();
+}
+
+bool MarkInvariantShapePath::run(lowered::LinearIR& linear_ir, lowered::LinearIR::constExprIt begin, lowered::LinearIR::constExprIt end) {
+    OV_ITT_SCOPED_TASK(ov::pass::itt::domains::SnippetsTransform, "Snippets::MarkInvariantShapePath");
+
+    size_t color_path = 0;
+
+    auto merge_paths = [&color_path](size_t lhs, size_t rhs) {
+        if (lhs == rhs || rhs == NOT_AFFECTING_PATH) return lhs;
+        if (lhs == NOT_AFFECTING_PATH) return rhs;
+        return ++color_path;
+    };
+
+    for (auto expr_it = begin; expr_it != end; ++expr_it) {
+        const auto& expr = *expr_it;
+        if (ov::is_type<ov::snippets::op::LoopBase>(expr->get_node()))
+            continue;
+
+        for (size_t out_idx = 0; out_idx < expr->get_output_count(); ++out_idx) {
+            size_t current_color_path;
+            if (is_elementwise_op(expr)) {
+                current_color_path = NOT_AFFECTING_PATH;
+                for (size_t in_idx = 0; in_idx < expr->get_input_count(); ++in_idx) {
+                    const auto input_path = getInvariantPortShapePath(expr->get_input_port(in_idx));
+                    current_color_path = merge_paths(current_color_path, input_path);
+                }
+            } else if (is_scalar_op(expr)) {
+                current_color_path = NOT_AFFECTING_PATH;
+            } else {
+                current_color_path = ++color_path;
```

Comment:
> Ideally, if we added a new operation and didn't update this pipeline, we should get an exception, so we wouldn't be able even to merge these changes.

Agree, but this pass is common pass and doesn't know about backend-specific ops (for example, `BrgemmCopyB`). And the pass will throw exception for backend-specific ops. I thought about using `ShapeInfer` instance of expressions (take from expr or use ShapeInferFactory from LIR) - but w don't have API to get `ShapeInfer` instances. So I had to use `if - else` with mention of expression/node types.

>  This way we wouldn't allocate excessive resources, but if a new color is really needed some tests will crash and we'll have to update this pass anyway.

Probably, it will work for the current ops (all expressions pass shape through except brgemm). When we implement new op, the test will be failed and we will need to update this pass (however, I hope that we will have new algorithm based on symbols). I hope debugging won't be expensive in this case.

Reply:
> Probably, it will work for the current ops (all expressions pass shape through except brgemm). When we implement new op, the test will be failed and we will need to update this pass

If we need this pass to work with backend-specific ops, we can always derive from this pass and override virtual checking functions like `is_not_affecting_op`, and then register the pass from the cpu node (probably disable generic pass first, then add backend customized version instead).
So it looks like the current design can easily be extended to backend-specific ops, if needed.